api/server/spanconverter.go

```diff
@@ -0,0 +1,129 @@
+package server
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+	"unicode"
+
+	"go.opencensus.io/stats"
+	view "go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+// SpanConverter registers as a opencensus Trace Exporter,
+// but it converts all the Spans in Views and registers them as such
+// A View exporter will then export them as normal.
+type SpanConverter struct {
+	opts     Options
+	measures map[string]*stats.Float64Measure
+	viewsMu  sync.Mutex
+	e        view.Exporter
+}
+
+// Options contains options for configuring the exporter.
+type Options struct {
+	Namespace string
+}
+
+func NewSpanConverter(o Options) (*SpanConverter, error) {
+	c := &SpanConverter{
+		opts:     o,
+		measures: make(map[string]*stats.Float64Measure),
+	}
+	return c, nil
+}
+
+var maxViews = 100
+
+// Spans are rejected if there are already maxViews (100) or they are
+// prefixed with '/', gin as been observed creating Span id specific
+// named Spans.
+func (c *SpanConverter) rejectSpan(sd *trace.SpanData) bool {
+	return len(c.measures) > maxViews || urlName(sd)
+}
+
+// ExportSpan creates a Measure and View once per Span.Name, registering
+// the View with the opencensus register. The length of time reported
+// by the span is then recorded using the measure.
+func (c *SpanConverter) ExportSpan(sd *trace.SpanData) {
+	if c.rejectSpan(sd) {
+		return
+	}
+	m := c.getMeasure(sd)
+
+	spanTimeNanos := sd.EndTime.Sub(sd.StartTime)
+	spanTimeMillis := float64(int64(spanTimeNanos / time.Millisecond))
+
+	stats.Record(context.Background(), m.M(spanTimeMillis))
+}
+
+func (c *SpanConverter) getMeasure(span *trace.SpanData) *stats.Float64Measure {
+	sig := sanitize(span.Name)
+	c.viewsMu.Lock()
+	m, ok := c.measures[sig]
+	c.viewsMu.Unlock()
+
+	if !ok {
+		m = stats.Float64(sig+"_span_time", "The span length in milliseconds", "ms")
+		v := &view.View{
```

Comment:
this seems like the wrong place for view registration. we should maybe/probably `stats.Record` everything and not limit the `stats.Record` bit (it's gated by `rejectSpan` as well as `getMeasure` and an arbitrary 100 limit imposed). at least, `stats.Record` and `view.Measure` are separate processes and should be exclusive. Tolga linked to some pointers for how the other stuff is handled atm, I think initially we're mostly registering views in initialization, we can maybe make this more friendly (ie no reboot required) later but it's relatively nice to work with for the ability to toggle them on/off for eg things that extend fn. then we end up whitelisting things in, which I kind of agree is painful, while at the same time it does make us organize things (like stat/trace names) so it's a net good I think. I think we can remove the limits (100) with that in place.

Reply:
(also won't require a lock - we shouldn't be locking for a stats library, that is on them!)

Reply:
Everything has already been recorded, that's already happened to produce the SpanData. This is copying up the recorded values into the Measure.